core/jvm/src/main/scala/scalaz/zio/lockfree/RingBuffer.scala

```diff
@@ -0,0 +1,147 @@
+package scalaz.zio.lockfree
+
+import java.util.concurrent.atomic.{ AtomicLong, AtomicReferenceArray }
+
+class RingBuffer[A](val desiredCapacity: Int) extends LockFreeQueue[A] {
+  final val capacity: Int                              = nextPow2(desiredCapacity)
+  private val buf: AtomicReferenceArray[BufferElement] = new AtomicReferenceArray[BufferElement](capacity)
```

Comment:
If we store anything extra beyond the `A`, it's not going to be very fast. So at most you can do: `AtomicReferenceArray[A]`. I know you're storing extra information in there, but we'll have to find ways around that. Also I'm not sure we gain from using `AtomicReferenceArray`, if all we're storing in there is an `A`, so I think we actually would prefer `Array[A]`, or possibly, `Array[AnyRef]`, so we erase the type.